File: prob1.py
# Time Complexity : O(n**2*!n)
# Space Complexity :O(n)
# Passed on Leetcode: yes

class Solution:
    def isSafe(self, r, c, arr):
        # col check
        for i in range(r):
            if arr[i][c] == True:
                return False
        
        # diag left check
        i = r
        j = c
        while i >= 0 and j >= 0:
            if arr[i][j] == True:
                return False
            i -= 1
            j -= 1
        
        # diag right check
        i = r
        j = c
        while i >= 0 and j < len(arr[0]):
            if arr[i][j] == True:
                return False
            i -= 1
            j += 1

        return True

    def backtrack(self, arr, i, n):
        # base
        if i == n:
            res = []
            for r in range(n):
                s = ''
                for c in range(n):
                    if arr[r][c] == False:
                        s += '.'
                    else:
                        s += 'Q'
                res.append(s)

            self.result.append(res)

        # logic
        for j in range(n):
            # action
            if self.isSafe(i, j, arr):
                arr[i][j] = True
                # recurse
                self.backtrack(arr, i+1, n)
                # backtrack
                arr[i][j] = False

    def solveNQueens(self, n: int) -> List[List[str]]:
        self.result = []
        self.arr = [[False for _ in range(n)] for _ in range(n)]
        self.backtrack(self.arr, 0, n)
        return self.result

File: prob1.java
// Time Complexity : O(n^2)
// Space Complexity : O(n!*n^2)
// Did this code successfully run on Leetcode : Yes


class Solution {
public:
    bool isSafe(int row, int col, vector<string>& board, int n)
    {
        //check for same row
        int x = row;
        int y = col;

        while(y >= 0)
        {
            if(board[x][y] == 'Q')
            {
                return false;
            }
            y--;
        }

        //check for upper diagonals

        x = row;
        y = col;

        while(x >= 0 && y >= 0)
        {
            if(board[x][y] == 'Q')
            {
                return false;
            }
            x--;
            y--;
        }

        x = row;
        y = col;

        while(x < n && y >= 0)
        {
            if(board[x][y] == 'Q')
            {
                return false;
            }
            x++;
            y--;
        }
        return true;
    }
    void solve(int col, vector<vector<string>>& ans, vector<string>& board, int n)
    {
        if(col == n)
        {
            ans.push_back(board);
            return;
        }

        //solve 1 case and rest Recursion will take care
        for(int row = 0; row < n; row++)
        {
            if(isSafe(row, col, board, n))
            {
                board[row][col] = 'Q';
                solve(col + 1, ans, board, n);
                //backtrack
                board[row][col] = '.';
            }
        }
    }
    vector<vector<string>> solveNQueens(int n) {
          vector<string> board(n, string(n, '.'));
          vector<vector<string>> ans;

          solve(0, ans, board, n);

          return ans;
    }
};